File: android-app/app/src/main/java/com/chainlesschain/android/pdh/social/toutiao/ToutiaoSignBridge.kt
package com.chainlesschain.android.pdh.social.toutiao

import android.content.Context
import com.chainlesschain.android.pdh.social.WebSignBridge
import dagger.hilt.android.qualifiers.ApplicationContext
import org.json.JSONObject
import javax.inject.Inject
import javax.inject.Singleton

/**
 * §A8 v0.3 — Toutiao `_signature` provider, runs ByteDance's `acrawler.js`
 * in a hidden WebView pre-warmed with the user's login cookie.
 *
 * Function-name rotation: ByteDance attaches the signing entry to one of
 * a small rotating set of globals; we probe them in order. Verified
 * 2026-05-24 against live www.toutiao.com:
 *
 *   - `window.byted_acrawler.sign({url, ...})` — current as of 2026-05;
 *     params object MUST contain `url`. Optional fields: `aid` (24 for
 *     Toutiao web), `token` (auth token from cookie).
 *   - `window._0x32d839(url)` — older naked-string variant, still observed
 *     on some user-agents.
 *   - `window.acrawler.sign(url)` — initial public name, kept as a last
 *     resort.
 *
 * Output: the signing function returns either a bare base64-ish string
 * (the `_signature` value alone) OR an object like `{_signature: "..."}`.
 * [buildSignScript] handles both — if the result is an object, we
 * `JSON.stringify` and let the bridge's [WebSignBridge.decodeJsResult] +
 * caller parse the `_signature` field out.
 */
@Singleton
class ToutiaoSignBridge @Inject constructor(
    @ApplicationContext context: Context,
) : WebSignBridge(context) {

    override val homepageUrl: String = "https://www.toutiao.com/"

    // Leading dot enables matching for www.toutiao.com / sf.toutiao.com /
    // www.ixigua.com etc. (Toutiao's acrawler is shared across the family).
    override val cookieDomain: String = ".toutiao.com"

    /**
     * `acrawler.js` finishes its init via a setTimeout chain after
     * `onPageFinished`. Real-device measurements 2026-05-24 (Xiaomi
     * 24115RA8EC): the global `byted_acrawler.sign` is observed
     * available 1100-1800ms after `onPageFinished`. 2500ms gives a
     * generous safety margin without sacrificing UX.
     */
    override val postLoadDelayMs: Long = 2500L

    override fun buildSignScript(rawUrl: String, purpose: String): String {
        // We pass the URL as the `url` field in an args object — this is
        // the only variant that's stable across all three function names.
        val args = JSONObject().apply {
            put("url", rawUrl)
            put("aid", AID_TOUTIAO_WEB)
            put("platform", "PC")
        }.toString()
        // Probe: bytes_acrawler.sign returns object, _0x32d839 + acrawler.sign
        // typically return string. Bridge's JSON.stringify wrapping covers
        // both. Caller [signUrl] override picks _signature out of the result.
        return """
            (function() {
              try {
                if (window.byted_acrawler && typeof window.byted_acrawler.sign === 'function') {
                  var r = window.byted_acrawler.sign($args);
                  if (typeof r === 'string') return r;
                  if (r && typeof r === 'object' && r._signature) return r._signature;
                  if (r && typeof r === 'object') return JSON.stringify(r);
                }
                if (typeof window._0x32d839 === 'function') {
                  return window._0x32d839(${JSONObject.quote(rawUrl)});
                }
                if (window.acrawler && typeof window.acrawler.sign === 'function') {
                  return window.acrawler.sign(${JSONObject.quote(rawUrl)});
                }
                return null;
              } catch (e) {
                return null;
              }
            })();
        """.trimIndent()
    }

    override suspend fun signUrl(rawUrl: okhttp3.HttpUrl, purpose: String): okhttp3.HttpUrl? {
        // Base impl returns the URL with the JS-output fragment appended
        // verbatim. For Toutiao, the JS returns just the `_signature` value
        // — we wrap it ourselves.
        val baseSigned = super.signUrl(rawUrl, purpose) ?: return null
        // If buildSignScript returned a bare value (not name=value), the base
        // impl will have appended it as a key with empty value. Detect that
        // and rebuild.
        val maybeSig = baseSigned.queryParameter("_signature")
        if (!maybeSig.isNullOrBlank()) return baseSigned
        // Fall back: extract the spurious bare-key we just appended, treat
        // the key itself as the signature, append properly. This handles
        // the `_0x32d839("https://...")` path returning just the base64.
        val spuriousKeys = mutableListOf<String>()
        for (i in 0 until baseSigned.querySize) {
            val name = baseSigned.queryParameterName(i)
            val value = baseSigned.queryParameterValue(i) ?: ""
            if (value.isEmpty() && name != "_signature" && rawUrl.queryParameter(name) == null) {
                spuriousKeys.add(name)
            }
        }
        if (spuriousKeys.size != 1) return baseSigned
        val sig = spuriousKeys[0]
        val rebuilt = rawUrl.newBuilder().addQueryParameter("_signature", sig).build()
        return rebuilt
    }

    companion object {
        const val AID_TOUTIAO_WEB = "24"

        /** Purpose tags passed to [signUrl] — kept as constants so callers can't typo. */
        const val PURPOSE_FEED = "feed"
        const val PURPOSE_COMMENTS = "comments"
        const val PURPOSE_SEARCH = "search"
    }
}

File: desktop-app-vue/src/main/sign-bridge/toutiao-sign-bridge.js
"use strict";

/**
 * Phase 6c (2026-05-25) — Toutiao `_signature` provider using Electron
 * WebContentsView.
 *
 * Node port of `android-app/.../pdh/social/toutiao/ToutiaoSignBridge.kt`.
 * Unlike Xhs (header signing), Toutiao mutates the URL by appending
 * `?_signature=<value>` — `signUrl` is the entrypoint, `signedHeaders`
 * stays as the base no-op.
 *
 * **Function-name rotation**: ByteDance attaches the signer to one of a
 * rotating set of globals; we probe in order. Verified 2026-05-24 against
 * live www.toutiao.com:
 *   1. `window.byted_acrawler.sign({url, aid, platform})` — current as of
 *      2026-05. Returns either a bare string OR `{_signature: "..."}`.
 *   2. `window._0x32d839(url)` — older naked-string variant, still seen on
 *      some user-agents.
 *   3. `window.acrawler.sign(url)` — initial public name, last-resort.
 *
 * **Output decode**: JS may return a bare base64-ish string (raw sig
 * value) OR an object `{_signature: "..."}`. We normalize to a URL with
 * `?_signature=<value>` query parameter appended.
 *
 * **Why URL mutation, not headers**: Toutiao's `_signature` is part of
 * the URL signature (encoded into query — server validates request URL
 * matches signature). Cannot be moved to headers without breaking the
 * validation.
 */

const { ElectronWebSignBridge } = require("./electron-web-sign-bridge");

class ToutiaoSignBridge extends ElectronWebSignBridge {
  constructor(opts = {}) {
    super(opts);
  }

  get homepageUrl() {
    return "https://www.toutiao.com/";
  }

  get cookieDomain() {
    // Leading dot enables matching for www.toutiao.com / sf.toutiao.com /
    // www.ixigua.com etc. (Toutiao's acrawler is shared across the family).
    return ".toutiao.com";
  }

  /**
   * acrawler.js finishes init via setTimeout chain after did-finish-load.
   * Android measurements (Xiaomi 24115RA8EC) showed 1100-1800ms; Electron
   * desktop is similar. 2500ms gives a safety margin without sacrificing
   * UX.
   */
  get postLoadDelayMs() {
    return 2500;
  }

  /**
   * Build JS probing 3 candidate globals (rotation history). All return
   * either a bare string (the sig value) OR an object containing
   * `_signature` field. We JSON.stringify objects so the bridge's eval
   * returns a string we can parse in [signUrl].
   *
   * @param {string} rawUrl  full URL string to sign
   * @param {string} _purpose  ignored — Toutiao signs the whole URL
   */
  buildSignScript(rawUrl, _purpose) {
    const urlJs = JSON.stringify(rawUrl);
    const argsJs = JSON.stringify({
      url: rawUrl,
      aid: AID_TOUTIAO_WEB,
      platform: "PC",
    });
    return `
      (function() {
        try {
          // Candidate 1: byted_acrawler.sign({url, aid, platform}) — 2026-05 current
          if (window.byted_acrawler && typeof window.byted_acrawler.sign === 'function') {
            var r = window.byted_acrawler.sign(${argsJs});
            if (typeof r === 'string') return r;
            if (r && typeof r === 'object' && r._signature) return r._signature;
            if (r && typeof r === 'object') return JSON.stringify(r);
          }
          // Candidate 2: _0x32d839(url) — older obfuscated name
          if (typeof window._0x32d839 === 'function') {
            return window._0x32d839(${urlJs});
          }
          // Candidate 3: acrawler.sign(url) — initial public name
          if (window.acrawler && typeof window.acrawler.sign === 'function') {
            return window.acrawler.sign(${urlJs});
          }
          return null;
        } catch (e) {
          return null;
        }
      })();
    `.trim();
  }

  /**
   * Sign the URL by appending `?_signature=<value>` (or `&_signature=...`
   * if existing query). Returns null on bridge cold / rotation / JS error.
   *
   * @param {URL|string} rawUrl
   * @param {string} purpose  passed through to [buildSignScript] (unused)
   * @returns {Promise<URL|null>}
   */
  async signUrl(rawUrl, purpose) {
    const urlStr = String(rawUrl);
    const result = await this._eval(urlStr, purpose);
    if (!result) {
      return null;
    }
    // Decode: either a bare string sig OR a JSON-encoded object.
    let signature = null;
    const trimmed = result.trim();
    if (trimmed.startsWith("{")) {
      try {
        const parsed = JSON.parse(trimmed);
        if (parsed && typeof parsed === "object" && parsed._signature) {
          signature = String(parsed._signature);
        }
      } catch {
        // Fall through — treat the whole string as the sig value.
      }
    }
    if (!signature) {
      // Treat as bare sig value. Filter obvious junk (empty / 'null' /
      // whitespace) so we don't append garbage to the URL.
      if (
        trimmed === "null" ||
        trimmed === "undefined" ||
        trimmed.length === 0
      ) {
        return null;
      }
      signature = trimmed;
    }
    // Append/replace _signature query param. URL handles encoding.
    const url = new URL(urlStr);
    url.searchParams.set("_signature", signature);
    return url;
  }

  /**
   * Toutiao signs URLs only — headers stay unchanged. Returns `{}`.
   */
  async signedHeaders(_rawUrl, _purpose) {
    return {};
  }

  /**
   * Phase 6e — return JSON-stringified candidate presence map for the
   * 3 known Toutiao signing globals.
   */
  get probeScript() {
    return `JSON.stringify({
      'byted_acrawler.sign': !!(window.byted_acrawler && typeof window.byted_acrawler.sign === 'function'),
      _0x32d839: typeof window._0x32d839 === 'function',
      'acrawler.sign': !!(window.acrawler && typeof window.acrawler.sign === 'function')
    });`;
  }
}

/** Toutiao web client id (Douyin web = 2906). */
const AID_TOUTIAO_WEB = "24";

module.exports = {
  ToutiaoSignBridge,
  // Exposed for unit testing
  _internals: {
    AID_TOUTIAO_WEB,
  },
};
